### PR TITLE
filter persons view

### PR DESCRIPTION
**Pull request by rerowep**
## Why are you opening this PR?

- https://tree.taiga.io/project/rero21-reroils/task/1188?kanban-status=1224894

## Code review check list

- [x] Commit message template compliance.
- [x] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?

### CODE REVIEW

File: rero_ils/modules/persons/api.py [hunk 1]
```diff
@@ -161,3 +163,18 @@ def _get_author_for_document(self):
...
+        results = DocumentsSearch().filter(
+            'term', authors__pid=self.pid
+        ).execute()
+        for result in results:
+            hold_pids = Holding.get_holdings_pid_by_document_pid(result['pid'])
+            for hold_pid in hold_pids:
+                organisations.append(
+                    Holding.get_record_by_pid(hold_pid).organisation_pid
+                )
+        return organisations
```
[the remark this answers is not in the document]
Comment (benerken) [reply rero_ils/modules/persons/api.py]: yes.... should be faster to use ES index to get the aggregation of organisations.

File: scripts/setup [hunk 2]
```diff
@@ -188,28 +188,28 @@ info_msg "Generate fixtures:"
...
 eval ${PREFIX} pipenv run invenio fixtures create --pid_type ptty ${DATA_PATH}/patron_types.json --append ${CREATE_LAZY} ${DONT_STOP}
-eval ${PREFIX} pipenv run invenio index reindex -t ptty --yes-i-know
+eval ${PREFIX} pipenv run invenio utils reindex -t ptty --yes-i-know
 
 info_msg "- Circulation policies: ${CREATE_LAZY} ${DONT_STOP}"
 eval ${PREFIX} pipenv run invenio fixtures create --pid_type cipo ${DATA_PATH}/circulation_policies.json --append ${CREATE_LAZY} ${DONT_STOP}
-eval ${PREFIX} pipenv run invenio index reindex -t cipo --yes-i-know
-eval ${PREFIX} pipenv run invenio index run --raise-on-error
+eval ${PREFIX} pipenv run invenio utils reindex -t cipo --yes-i-know
+eval ${PREFIX} pipenv run  invenio utils runindex --raise-on-error
```
Comment (blankoworld) [inline scripts/setup]: Why two spaces between run and invenio words?

File: rero_ils/modules/persons/api.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rerowep) [reply rero_ils/modules/persons/api.py]: Ok implemented

File: scripts/setup (hunk 2 above)
Comment (rerowep) [reply scripts/setup]: Replacement error

File: rero_ils/modules/persons/api.py [hunk 3]
```diff
@@ -161,3 +161,20 @@ def _get_author_for_document(self):
...
+                field='holdings.organisation.organisation_pid', size=10)
+        search.aggs.bucket('organisation', agg)
+        results = search.execute()
+        for result in results:
+            hold_pids = Holding.get_holdings_pid_by_document_pid(result['pid'])
+            for hold_pid in hold_pids:
+                organisations.append(
+                    Holding.get_record_by_pid(hold_pid).organisation_pid
+                )
+        return list(set(organisations))
```
Comment (lauren-d) [inline rero_ils/modules/persons/api.py]: why you doesn't use the result of aggregation to get organisation pid ?
Something like :
```python
results = search.execute()
for result in results.aggregations.organisation.buckets :
  if result.doc_count > 0 :
    organisations.append(result.key)
return organisations
```

Comment (rerowep) [reply rero_ils/modules/persons/api.py]: Because I did not know how to do it.

File: rero_ils/modules/documents/api.py [hunk 4]
```diff
@@ -45,6 +47,22 @@
 document_id_fetcher = partial(id_fetcher, provider=DocumentProvider)
 
 
+class DocumentsIndexer(IlsRecordIndexer):
+    """Indexing documents in Elasticsearch."""
+
+    def index(self, record):
+        """Index an document."""
+        return_value = super(DocumentsIndexer, self).index(record)
+        from ..persons.api import Person
```
Comment (jma) [inline rero_ils/modules/documents/api.py]: can you put this at the top of the file?

File: rero_ils/modules/documents/api.py [hunk 5]
```diff
@@ -45,6 +47,22 @@
...
+        for author in record.replace_refs().get('authors', []):
+            person = Person.get_record_by_pid(author.get('pid'))
+            if person:
+                person.reindex()
+        index_name, doc_type = current_record_to_index(record)
+        current_search.flush_and_refresh(index_name)
```
Comment (jma) [inline rero_ils/modules/documents/api.py]: This is called twice (here and during the super() call.)

File: rero_ils/modules/persons/api.py [hunk 6]
```diff
@@ -161,3 +160,17 @@ def _get_author_for_document(self):
...
+        organisations = set()
+        search = DocumentsSearch().filter('term', authors__pid=self.pid)
+        agg = A('terms',
+                field='holdings.organisation.organisation_pid', size=10)
+        search.aggs.bucket('organisation', agg)
+        results = search.execute()
+        for result in results.aggregations.organisation.buckets:
+            if result.doc_count:
+                organisations.add(result.key)
+        return sorted(list(organisations))
```
Comment (jma) [inline rero_ils/modules/persons/api.py]: does it make sense to sort the results?

File: rero_ils/modules/documents/api.py (hunk 4 above)
Comment (rerowep) [reply rero_ils/modules/documents/api.py]: Nop, circular inclusion!

File: rero_ils/modules/persons/api.py (hunk 6 above)
Comment (rerowep) [reply rero_ils/modules/persons/api.py]: No, changed!

File: rero_ils/modules/documents/api.py (hunk 5 above)
Comment (jma) [reply rero_ils/modules/documents/api.py]: not addressed

File: rero_ils/modules/documents/api.py [hunk 7]
```diff
@@ -45,6 +47,22 @@
...
 
 
+class DocumentsIndexer(IlsRecordIndexer):
+    """Indexing documents in Elasticsearch."""
+
+    def index(self, record):
+        """Index an document."""
+        return_value = super(DocumentsIndexer, self).index(record)
+        from ..persons.api import Person
+        for author in record.replace_refs().get('authors', []):
```
Comment (BadrAly) [inline rero_ils/modules/documents/api.py]: An idea: Can the index of authors become optional (default=Y) ? I am thinking when there is a mass correction at the document level that does not touch the authorities. This way we can bypass the authority indexing.

Comment (rerowep) [reply rero_ils/modules/documents/api.py]: Mass corrections can use bulk indexing.

**Pull request merged**